### ISSUE TITLE
Parallel execution of job groups on SLURM

### ISSUE THREAD

**Issue by koen-vg** [snakemake/snakemake#2060]
**Snakemake version**
7.19.1

**Describe the bug**
I'm sure this is on the agenda, but since e6cdb32fd20a9899f441b3ec8aca3f36710f7a4a, job groups are run entirely sequentially on SLURM even if multiple jobs in the group could be run in parallel. Even so, the runtime for the whole group is calculated with the assumption that jobs that can be run in parallel are run in parallel. This means that job groups can easily time out even though the runtime specification for each individual rule is entirely correct.

**Minimal example**
Something like
```
rule all: [...]

rule a:
    ...
    resources: runtime=1
    group: "foo"
    shell: "sleep 0.8"

rule b:
    ...
    resources: runtime=1
    group: "foo"
    shell: "sleep 0.8"
```
called with `snakemake --slurm -j2 -c2 all` will trigger the problem.

**Additional context**
Please let me know if I can get started in solving this in any way! I'm more than happy to contribute, and I'm starting to know my way around the snakemake codebase a little bit.

**Comment by pvandyken** [comment]
I'm not so familiar with the slurm executor code (@cmeesters is who we want), but I did write the runtime resource calculation for group resources.

My question is why are slurm jobs not being run in parallel? Is there a decisive reason, or is it an oversight or mistake?

**Comment by cmeesters** [comment]
I'm afraid, but I cannot reproduce this behaviour. To confirm, I changed the `test_group_job` Snakefile to contain `touch {output} && sleep 300` to give me enough time. I submitted with `snakemake -j unlimited -c 4 --slurm --default-resources ...`, logged in the job node and:

```
z0133:~ meesters$ ps aux | grep meesters | grep sleep
meesters  142081  0.0  0.0 222836  3476 ?        S    18:57   0:00 /usr/bin/bash -c set -euo pipefail;  touch a/2.out && sleep 300
meesters  142083  0.0  0.0 217144   976 ?        S    18:57   0:00 sleep 300
meesters  142486  0.0  0.0 222836  3376 ?        S    18:57   0:00 /usr/bin/bash -c set -euo pipefail;  touch a/5.out && sleep 300
meesters  142488  0.0  0.0 217144   968 ?        S    18:57   0:00 sleep 300
meesters  142591  0.0  0.0 222836  3480 ?        S    18:57   0:00 /usr/bin/bash -c set -euo pipefail;  touch a/1.out && sleep 300
meesters  142593  0.0  0.0 217144   896 ?        S    18:57   0:00 sleep 300
meesters  142608  0.0  0.0 221932  1144 pts/0    S+   18:58   0:00 grep --color=auto sleep
``` 

Exactly, what we are supposed to see. 

I am not sure what you mean @pvandyken : snakemake's SLURM jobstep executor has no build-in parallelism. It rather triggers its job step sequentially, such that the slurmstepd will ensure _concurrent_ execution. 

Likewise for individual jobs: They are submitted one by one, the scheduler will attribute resources accordingly. snakemake itself is in this context not able to decide about concurrency, except for `localrules` under regime of the CPUExecutor, of course. The number of launched jobs is throttled by the `-j` semaphore. 

As you mention a particular commit, @koen-vg , may I assume that you are working with a development version? I too, am a little confused about this change as expressed in the discussion to PR #1861 . I did not realize, the change is present in the mentioned commit. So: I am not sure about which code we are talking here, and _this_ is important. The commit erased (commented out) the group-part in the Executor entirely. Without it, the functionality is gone. @johanneskoester could you elaborate on this finding? Am I overlooking something here?

**Comment by cmeesters** [comment]
Arrgh, I stand corrected: The outcommented code made to the release! But why? 

I will look into the matter, a.s.a.p.: Alas, I will have little time, tomorrow.

**Comment by pvandyken** [comment]
Sounds good. Just to clarify terminology, I think I've been using parallelism in the same sense as you're using concurrent, which is probably imprecise in my part.

**Comment by cmeesters** [comment]
It must have been a wrong merge, the code looks like one from a previous discussion. The latest code in the `SLURM_Executor`-branch is fine and up to date. Johannes (@johanneskoester ), I think something has been a tiny bit messed up. ;-)

**Comment by cmeesters** [comment]
> Sounds good. Just to clarify terminology, I think I've been using parallelism in the same sense as you're using concurrent, which is probably imprecise in my part.

It's a professional deformation of mine: I am teaching the distinction too often. ;-)

**Comment by johanneskoester** [comment]
@cmeesters mhm, sorry for not specifying the exact reason for that change. It happened [here](https://github.com/snakemake/snakemake/pull/2006), and I did this to fix a problem with that step wise group job execution. Unfortunately I don't exactly remember. It would be great if you could just open a PR that adds it in again, and we will hopefully just see the issue in the CI then (if there still is one).

**Comment by johanneskoester** [comment]
Nevertheless, even the current implementation should still run the jobs inside a group in parallel. If it doesn't we probably don't pass the correct number of cores to the srun argument in the SMP variant of the jobstep execution.

**Comment by cmeesters** [comment]
@johanneskoester "Nevertheless, even the current implementation should still run the jobs inside a group in parallel (as Snakemake can do that itself without slurm)" - not when invoked with `--slurm` as it then calls for the slurm_jobstep-executor, which - if outcommented - does not support group jobs.

Btw. the SLURM version mentioned in the linked commit is so old, that supporting it would never occur to me: There are a number of security fixes, which need attention. Anything significantly older than 20.11 should not be considered for support (at the time of writing this comment). 

However, I will do a PR a.s.a.p.-ish (tomorrow?).

**Comment by johanneskoester** [comment]
> not when invoked with --slurm as it then calls for the slurm_jobstep-executor, which - if outcommented - does not support group jobs.

I don't really see why. The jobstep executor still gets the group job, and if there is no special casing for group jobs, it will consider it to be an ordinary SMP job and execute Snakemake to run it. Snakemake will then schedule the individual jobs of the group job locally. If this happens to be not parallelized, then there is just an issue in the way we invoke Snakemake (e.g. perhaps the resources are somehow not constrained in the right way).

**Comment by johanneskoester** [comment]
I've just tested it locally, and it works fine here, groups jobs are executed in parallel if the available nodes have enough CPU cores, and their requirements are also correctly propagated. Here is a PR that tries to perform this testcase in our CI. I hope that the slurm cluster there has enough cores (needs 2 for the parallelism): https://github.com/snakemake/snakemake/pull/2092

**Comment by johanneskoester** [comment]
THe test case passes. Hence, I would consider this to be fine. However, please feel free to provide a counter example, ideally via a PR with an additional testcase.

**Comment by johanneskoester** [comment]
Please reopen or post a new issue if you see this again. Also notify me via discord, as I might miss the github notification.

**Comment by gipert** [comment]
Hi, it seems to me this issue is not resolved. I tried with the following Snakefile:
```
rule all:
    input: "a.dat", "b.dat"

rule a:
    output: "a.dat"
    resources: runtime=10
    group: "foo"
    shell: "touch {output} && sleep 200"

rule b:
    output: "b.dat"
    resources: runtime=10
    group: "foo"
    shell: "touch {output} && sleep 200"
```

And submitted jobs at NERSC via SLURM as follows (ignore the missing partition warning, not relevant at NERSC):

```console
> snakemake --slurm --default-resources slurm_account="m2676" constraint="cpu" 'slurm_extra="--qos regular"' -j unlimited -c all --group-components 'foo=2' -- all
Building DAG of jobs...
Using shell: /usr/bin/bash
Provided cores: 256
Rules claiming more threads will be scaled down.
Job stats:
job      count    min threads    max threads
-----  -------  -------------  -------------
a            1              1              1
all          1              1              1
b            1              1              1
total        3              1              1

Select jobs to execute...
[Tue Jul 11 01:52:07 2023]

group job foo (jobs in lexicogr. order):

    [Tue Jul 11 01:52:07 2023]
    rule a:
        output: a.dat
        jobid: 1
        reason: Missing output files: a.dat
        resources: mem_mb=1000, mem_mib=954, disk_mb=1000, disk_mib=954, tmpdir=<TBD>, slurm_account=m2676, constraint=cpu, slurm_extra=--qos regular, runtime=10


    [Tue Jul 11 01:52:07 2023]
    rule b:
        output: b.dat
        jobid: 2
        reason: Missing output files: b.dat
        resources: mem_mb=1000, mem_mib=954, disk_mb=1000, disk_mib=954, tmpdir=<TBD>, slurm_account=m2676, constraint=cpu, slurm_extra=--qos regular, runtime=10

No partition was given for rule 'JobGroup(foo,frozenset({a, b}))', and unable to find a default partition. Trying to submit without partition information. You may want to invoke snakemake with --default-resources 'slurm_partition=<your default partition>'.
Job da047075-65ce-5e6e-a797-7a8a471e2c14 has been submitted with SLURM jobid 11460682 (log: .snakemake/slurm_logs/group_foo/11460682.log).
[Tue Jul 11 02:03:17 2023]
Finished job 1.
[Tue Jul 11 02:03:17 2023]
Finished job 2.
2 of 3 steps (67%) done
Select jobs to execute...

[Tue Jul 11 02:03:17 2023]
localrule all:
    input: a.dat, b.dat
    jobid: 0
    reason: Input files updated by another job: b.dat, a.dat
    resources: mem_mb=1000, mem_mib=954, disk_mb=1000, disk_mib=954, tmpdir=/tmp, slurm_account=m2676, constraint=cpu, slurm_extra=--qos regular

[Tue Jul 11 02:03:17 2023]
Finished job 0.
3 of 3 steps (100%) done
Complete log: .snakemake/log/2023-07-11T015206.690822.snakemake.log
```

As you can see from the logs below, the two jobs are not executed concurrently. Each of them is assigned a separate `srun` call, which are run sequentially.

```console
> cat .snakemake/slurm_logs/group_foo/11460682.log
Building DAG of jobs...
Using shell: /usr/bin/bash
Provided cores: 256
Rules claiming more threads will be scaled down.
Provided resources: disk_mib=1908, mem_mib=1908, mem_mb=2000, disk_mb=2000
Select jobs to execute...
Building DAG of jobs...
Using shell: /usr/bin/bash
Provided cores: 2
Rules claiming more threads will be scaled down.
Provided resources: disk_mb=1000, mem_mib=954, disk_mib=954, mem_mb=1000
Select jobs to execute...

[Tue Jul 11 01:55:50 2023]
rule a:
    output: a.dat
    jobid: 0
    reason: Missing output files: a.dat
    resources: mem_mb=1000, mem_mib=954, disk_mb=1000, disk_mib=954, tmpdir=/tmp, slurm_account=m2676, constraint=cpu, slurm_extra=--qos regular, runtime=10

[Tue Jul 11 01:59:10 2023]
Finished job 0.
1 of 1 steps (100%) done
[Tue Jul 11 01:59:11 2023]
Finished job 0.
1 of 2 steps (50%) done
Select jobs to execute...
Building DAG of jobs...
Using shell: /usr/bin/bash
Provided cores: 2
Rules claiming more threads will be scaled down.
Provided resources: disk_mb=1000, mem_mib=954, disk_mib=954, mem_mb=1000
Select jobs to execute...

[Tue Jul 11 01:59:19 2023]
rule b:
    output: b.dat
    jobid: 0
    reason: Missing output files: b.dat
    resources: mem_mb=1000, mem_mib=954, disk_mb=1000, disk_mib=954, tmpdir=/tmp, slurm_account=m2676, constraint=cpu, slurm_extra=--qos regular, runtime=10

[Tue Jul 11 02:02:39 2023]
Finished job 0.
1 of 1 steps (100%) done
[Tue Jul 11 02:02:40 2023]
Finished job 1.
2 of 2 steps (100%) done
```
Related: #2339

**Comment by dlaehnemann** [comment]
Job grouping will only jointly submit rule instances that form a connected subgraph of the overall workflow DAG -- these are usually a number of rules that come one after the other in the workflow, and one connected subgraph is usually defined by a specific value for a wildcard. I think the example in this docs section does a good job at explaining this:
https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#snakefiles-grouping

If you want to group multiple parallel connected subgraphs of your rule DAG (i.e. group multiple wildcardvalues / samples) into one cluster job, `--group-components` is probably what you are looking for:
https://snakemake.readthedocs.io/en/stable/executing/grouping.html

**Comment by gipert** [comment]
I understand this, and I am indeed using that option, check out the command line in my second code block.

**Comment by dlaehnemann** [comment]
But you are not using a wildcard, so there is nothing that `--group-components` can run in parallel. Maybe playing around with the minimal example from the docs (and combining it with the `--group-components` option) can make this clearer:
https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#snakefiles-grouping

These are really weird concepts to get your head around. I guess some sort of visualization of the DAG might help improve the docs.

**Comment by gipert** [comment]
Sorry for the silence. I have tried again with this Snakefile:
```
rule all:
    input: "a.dat", "b.dat"

rule a:
    output: "{sample}.dat"
    resources: runtime=10
    group: "foo"
    shell: "touch {output} && sleep 200"
```
but jobs keep getting executed sequentially, exactly as before:
```
> cat .snakemake/slurm_logs/group_foo/13625118.log             pertoldi@login40
Building DAG of jobs...
Using shell: /usr/bin/bash
Provided cores: 256
Rules claiming more threads will be scaled down.
Provided resources: mem_mib=1908, mem_mb=2000, disk_mib=1908, disk_mb=2000
Select jobs to execute...
Building DAG of jobs...
Using shell: /usr/bin/bash
Provided cores: 2
Rules claiming more threads will be scaled down.
Provided resources: mem_mb=1000, mem_mib=954, disk_mb=1000, disk_mib=954
Select jobs to execute...

[Wed Aug  9 03:57:34 2023]
rule a:
    output: a.dat
    jobid: 0
    reason: Missing output files: a.dat
    wildcards: sample=a
    resources: mem_mb=1000, mem_mib=954, disk_mb=1000, disk_mib=954, tmpdir=/tmp, slurm_account=m2676, constraint=cpu, slurm_extra=--qos regular, runtime=50

srun: Job 13625118 step creation temporarily disabled, retrying (Requested nodes are busy)
[Wed Aug  9 04:00:54 2023]
Finished job 0.
1 of 1 steps (100%) done
srun: Job 13625118 step creation still disabled, retrying (Requested nodes are busy)
srun: Step created for StepId=13625118.1
Building DAG of jobs...
Nothing to be done (all requested files are present and up to date).
[Wed Aug  9 04:01:14 2023]
Finished job 0.
1 of 2 steps (50%) done
Select jobs to execute...
Building DAG of jobs...
Using shell: /usr/bin/bash
Provided cores: 2
Rules claiming more threads will be scaled down.
Provided resources: disk_mib=954, mem_mb=1000, disk_mb=1000, mem_mib=954
Select jobs to execute...

[Wed Aug  9 04:01:22 2023]
rule a:
    output: b.dat
    jobid: 0
    reason: Missing output files: b.dat
    wildcards: sample=b
    resources: mem_mb=1000, mem_mib=954, disk_mb=1000, disk_mib=954, tmpdir=/tmp, slurm_account=m2676, constraint=cpu, slurm_extra=--qos regular, runtime=50

srun: Job 13625118 step creation temporarily disabled, retrying (Requested nodes are busy)
[Wed Aug  9 04:04:42 2023]
Finished job 0.
1 of 1 steps (100%) done
srun: Job 13625118 step creation still disabled, retrying (Requested nodes are busy)
srun: Step created for StepId=13625118.3
Building DAG of jobs...
Nothing to be done (all requested files are present and up to date).
[Wed Aug  9 04:04:51 2023]
Finished job 1.
2 of 2 steps (100%) done
```
I have also added an additional print statement in the Snakemake source code to show me the actual sub-commands being executed on the compute node:
```
srun -n1 --cpu-bind=q /global/cfs/cdirs/m2676/users/pertoldi/.conda/snakemake/bin/python3.11 -m snakemake --snakefile '/global/cfs/cdirs/m2676/users/pertoldi/smk-mwe/Snakefile' --target-jobs 'a:sample=a' --allowed-rules 'a' --local-groupid 'bcdf3954-30da-5bc4-856d-5d7aa0401dcb' --cores 'all' --attempt 1 --resources 'disk_mib=954' 'mem_mb=1000' 'disk_mb=1000' 'mem_mib=954' --wait-for-files '/global/cfs/cdirs/m2676/users/pertoldi/smk-mwe/.snakemake/tmp.vadjpudv' --force --keep-target-files --keep-remote --max-inventory-time 0 --nocolor --notemp --no-hooks --nolock --ignore-incomplete --rerun-triggers 'mtime' 'code' 'input' 'params' 'software-env' --conda-frontend 'mamba' --wrapper-prefix 'https://github.com/snakemake/snakemake-wrappers/raw/' --latency-wait 5 --scheduler 'ilp' --scheduler-solver-path '/global/cfs/cdirs/m2676/users/pertoldi/.conda/snakemake/bin' --default-resources 'mem_mb=max(2*input.size_mb, 1000)' 'disk_mb=max(2*input.size_mb, 1000)' 'tmpdir=system_tmpdir' 'slurm_account=m2676' 'constraint=cpu' 'slurm_extra="--qos regular"' --mode 2
srun -n1 --cpu-bind=q /global/cfs/cdirs/m2676/users/pertoldi/.conda/snakemake/bin/python3.11 -m snakemake --snakefile '/global/cfs/cdirs/m2676/users/pertoldi/smk-mwe/Snakefile' --target-jobs 'a:sample=b' --allowed-rules 'a' --local-groupid '59e1e836-2806-5c78-8501-91af2d27d59b' --cores 'all' --attempt 1 --resources 'disk_mib=954' 'mem_mb=1000' 'disk_mb=1000' 'mem_mib=954' --wait-for-files '/global/cfs/cdirs/m2676/users/pertoldi/smk-mwe/.snakemake/tmp.vadjpudv' --force --keep-target-files --keep-remote --max-inventory-time 0 --nocolor --notemp --no-hooks --nolock --ignore-incomplete --rerun-triggers 'mtime' 'code' 'input' 'params' 'software-env' --conda-frontend 'mamba' --wrapper-prefix 'https://github.com/snakemake/snakemake-wrappers/raw/' --latency-wait 5 --scheduler 'ilp' --scheduler-solver-path '/global/cfs/cdirs/m2676/users/pertoldi/.conda/snakemake/bin' --default-resources 'mem_mb=max(2*input.size_mb, 1000)' 'disk_mb=max(2*input.size_mb, 1000)' 'tmpdir=system_tmpdir' 'slurm_account=m2676' 'constraint=cpu' 'slurm_extra="--qos regular"' --mode 2
```
